### PR TITLE
Performance

### PR DESCRIPTION
**Pull request by ricksbrown**
- Begins to address #263. There are definitely more optimizations to make but this is already a large PR.
- Addresses a TODO, namely expanding AX_ARIA_02 so that it covers HTML attributes as well as ARIA attributes.
- Fixed some style nits while I was there

Hopefully the performance improvement works on other pages (and hopefully you are happy with the changes).

Here are the results on my test page, using `console.time` (no audits disabled, includes DuplicateID):

Before: 7 seconds
After: 2 seconds (70% faster)

@alice PTAL and sorry it's a big one to review! There's about 40 hours work in it.

### CODE REVIEW

File: src/js/Audit.js [hunk 1]
```diff
@@ -218,68 +218,122 @@ axs.Audit.getRulesCannotRun = function(opt_configuration) {
...
+            level: 0,
+            ignoring: {},
+            disabled: false,
+            hidden: false
+        };
+        var root = configuration.walkDom ? document.documentElement : configuration.scope;
+        // Because 'related elements' could occur anywhere in the DOM we need to start at document.documentElement
+        axs.dom.composedTreeSearch(root, null, { preorder: function(element, flags) {
+            if (!flags.inScope)
+                flags.inScope = element === configuration.scope;
```
Comment (alice) [inline src/js/Audit.js]: This will need to be undone when leaving the scope - you could add a `postorder` function to achieve this.

File: src/js/Audit.js [hunk 2]
```diff
@@ -218,68 +218,122 @@ axs.Audit.getRulesCannotRun = function(opt_configuration) {
...
-    return results;
-};
-goog.exportSymbol('axs.Audit.run', axs.Audit.run);
-
+    /**
+     * Determine if this element should be ignored by the AuditRule.
+     * @param {Element} element An audit candidate.
+     * @returns {boolean} true if this element should be ignored by this AuditRule.
+     */
+    AuditOptions.prototype.ignore = function(element) {
```
Comment (alice) [inline src/js/Audit.js]: "ignore" here reads to me like the imperative; I would expect this to update the configuration to ignore the given element, based on the name.

What about something like `shouldIgnore` or `ignored` or `isIgnored`?

File: src/js/AuditRule.js [hunk 3]
```diff
@@ -47,13 +48,26 @@ axs.AuditRule = function(spec) {
     /**
      * Is this element relevant to this audit?
      * @param {Element} element A potential audit candidate.
+     * @param {Object} flags
```
Comment (alice) [inline src/js/AuditRule.js]: As I read it, `flags` is actually optional. If this is the case, let's annotate it as such (`{Object=}`) and name accordingly to be consistent with other optional params.

File: src/js/AccessibilityUtils.js [hunk 4]
```diff
@@ -549,10 +549,12 @@ axs.utils.isNativelyDisableable = function(element) {
  * This element may or may not be effectively disabled in practice as this is dependent on implementation.
  *
  * @param {Element} element An element to check.
+ * @param {boolean} self If true only consider the element itself.
```
Comment (alice) [inline src/js/AccessibilityUtils.js]: This param name is a little confusing, given `self` is often used to alias `this`. Perhaps something like `checkForInherited` (sense inverted)?

Also, this only limits the scope for the `aria-disabled` check, and doesn't prevent walking up the parent chain for native `disabled`. Is that deliberate? Perhaps early-out the for loop if this is true (or false if the sense is inverted, obviously)?

File: src/audits/DuplicateId.js [hunk 5]
```diff
@@ -25,15 +25,36 @@ axs.AuditRules.addRule({
...
         }
+        if (element.hasAttribute('id')) {
+            return true;
+        }
         return false;
     },
+    /**
+     * @this {axs.AuditRule}
+     */
+    isRelevant: function(element, flags) {
```
Comment (alice) [inline src/audits/DuplicateId.js]: This is pretty hard to follow.

It seems like a bunch of infrastructure in this change is simply to make this work. I think it's definitely worthwhile, but it'll be much easier to review if you can split it into two or three changes - potentially one with just the formatting fixes (thanks for that btw), then one with the main performance optimisation (the single pass for collectMatchingElements, and related changes), and then one for this extra optimisation.

I know from experience how annoying this is to do, but git makes it tractable, and it'll make the review a lot easier. Feel free to ping me on hangouts if you need any help with the mechanics - not sure how much of this git surgery you've had to do in the past, but I've definitely done my fair share and happy to share what I've learnt!

File: src/js/AuditRule.js [hunk 6]
```diff
@@ -47,13 +48,26 @@ axs.AuditRule = function(spec) {
...
      * @return {boolean} true if this element is relevant to this audit.
      */
     this.relevantElementMatcher_ = spec.relevantElementMatcher;
 
     /**
+     * If the relevantElementMatcher is not enough to determine whether or not this rule should run
+     *    then this function can be called (after DOM traversal is finished).
```
Comment (alice) [inline src/js/AuditRule.js]: Extra space before "then" isn't necessary - did the formatter add that?

**Comment by alice** [comment]
On first glance, this looks really solid and I'm really excited about getting it in, but as you noted it's a lot to bite off in one review. How would you feel about breaking it into three changes as described in the comment above?

File: src/audits/DuplicateId.js (hunk 5 above)
Comment (ricksbrown) [inline src/audits/DuplicateId.js]: Yep it's hard to follow, I should have commented it more and I'm glad you're looking at this while it's still in my head.

I can split it up, sure thing.
If you're in my neighbourhood feel free to come to my place and we could go over it together. It really does need explaining.

This particular line `isRelevant` is necessary to faithfully preserve the `DuplicateId` audit's behaviour:

- It needs to return `NA` if there are no relevant elements.
- Relevant elements are those which have one or more "id referrers".

The problem: ID referrers may be encountered **after** the element, so we have to wait until the DOM traversal is totally finished before we can know if there are *really* relevant elements.

Without `isRelevant` DuplicateId would return `PASS` instead of `NA` for every element with an ID.

File: src/js/Audit.js (hunk 1 above)
Comment (ricksbrown) [inline src/js/Audit.js]: I may have stuffed this up but the way it's meant to work is that every level of the DOM tree actually has its own copy of the `flags` object which inherits from the parent flags object (using the prototype chain).

This happens in `composedTreeSearch` (sorry to taint your masterpiece) on this line:
`var flags = Object.create(parentFlags);`

So `flags.inScope` only affects the current level of DOM traversal and its entire subtree.
Well that's the theory anyway...

Comment (alice) [inline src/js/Audit.js]: Got it. Will take another look at this once you've managed to factor out the three separate changes.

File: src/js/AuditRule.js (hunk 6 above)
Comment (ricksbrown) [inline src/js/AuditRule.js]: lol no - one Alice Boxhall [told me to do that](https://github.com/GoogleChrome/accessibility-developer-tools/pull/100#discussion_r16298003) :grinning:

File: src/js/Audit.js (hunk 2 above)
Comment (ricksbrown) [inline src/js/Audit.js]: Agreed - will be in the bite-sized PR.

File: src/js/AuditRule.js (hunk 3 above)
Comment (ricksbrown) [inline src/js/AuditRule.js]: It will always be there, I think it isn't optional? It will be easier to see when I give you the smaller PR.

File: src/js/AuditRule.js (hunk 6 above)
Comment (alice) [inline src/js/AuditRule.js]: Ah - that's only for params and returns :)

Comment (ricksbrown) [inline src/js/AuditRule.js]: Ah right, got it, thanks!

**Comment by ricksbrown** [comment]
@alice Hope this is OK, I've made it easier to pull parts out by splitting into three distinct commits.
Hopefully my git surgery was OK, thanks for the offer of help but I didn't want you to get frustrated at my git noobery :blush: 

We started with a PR that had three separate changes, 56 files changed, `+2,893 -2,006`

**Change 1, Stylistic Tweaks**
All gone (for now).

**Change 2, Audit Runner**
PART 1
The [core changes](https://github.com/ricksbrown/accessibility-developer-tools/commit/ef5cef634fc55423124ab0fb2001340a8e3c38a1) to the audit runner and a couple of tweaked AuditRules now account for: 
39 files changed  `+2,067 −1,157`. So that's literally thousands of changes you do not need to review, yay!

This does not deliver any significant performance boosts because the majority of the AuditRules are left untouched.

PART 2
[Updating the audit](https://github.com/ricksbrown/accessibility-developer-tools/commit/c282f927a81cdac0802a802dcd04526c5c6edbb0) rules adds:
12 changed files (eek!) but only `+31 -34` (phew!) 
These changes have a large impact impact on performance.

**Change 3, NonExistentRelatedElement**
I know you wanted a separate PR but it seemed very small: `+27 -38` so [I've thrown it in there](https://github.com/ricksbrown/accessibility-developer-tools/commit/272613e9f68398ec42fa8bf3ee7ad9c7dcfdd8aa).

Let me know if you still want me to remove this.

**Summary**
The grand total now stands at:
52 files changed `+2,118 −1,222`

Honestly the vast majority of changes are robotic cut and paste to the unit tests. Once you are over the complex core changes it should be smooth sailing.

I'd be happy to tune into hangouts or something and virtually sit with you while you go over it if you think it would help.

**Comment by alice** [comment]
Hey - sorry to be this person, but what did you think of my suggestion about splitting this up even further?

**Comment by ricksbrown** [comment]
As [discussed](https://web-a11y.slack.com/archives/a11y-devtools/p1450227033000083) git surgery pending - please hold off reviewing this until I post back here.

**Comment by ricksbrown** [comment]
@alice git surgery complete - it is now split into seven commits which, if you review the changes in order, should hopefully make more sense.

The first commit is the biggest because it changes the unit tests.

I realised I did make some stylistic changes by encapsulating certain functionality into new "classes" like `AuditOptions` or new helpers like `axs.AuditRules.getActiveRules`. The functionality used to be there and I have essentially shuffled it around. If you hate it encapsulated that way I could probably undo it to some extent. I didn't give it a huge amount of thought, I was just trying to break things down, reduce cyclomatic complexity and scope so I could get my head around it and see what was used where.

**Comment by alice** [comment]
Thank you SO MUCH for doing that git surgery. It made it so much easier to review and to see each idea in isolation. I really love this work - obviously I had a bunch of comments but no show stoppers; I'm really excited to get this in.

**Comment by addyosmani** [comment]
@ricksbrown This work looks 🏆. Do you think you'll have time to address any of the comments from @alice above? If not, it might be worth rebasing against master so they can be looked at in another round of work in the future.

**Comment by ricksbrown** [comment]
@addyosmani Thanks for the encouragement 😄 I dropped this for a while. I shall take a look at it pronto.

**Comment by addyosmani** [comment]
@alice were there any other pieces of feedback you wanted addressed before we landed this?

**Comment by addyosmani** [comment]
Currently have some CI failures for this PR I'm investigating. Going to restart it for this PR and see if we can go green.

**Comment by addyosmani** [comment]
Landing these changes now. We can address further feedback in a follow-up PR as time allows.

File: src/js/AuditRule.js (hunk 3 above)
Comment (dran7505) [inline src/js/AuditRule.js]: This diff
